File: 06-arrays-sorting/02_buy_sell_stock.py
import sys, os; sys.path.append(os.path.join(os.path.dirname(os.path.abspath(__file__)), '..'))
from test_runner import run_tests


# Best Time to Buy and Sell Stock (LC #121) -- Easy
# Max profit from single buy and sell.
#
#   Key: Track min price so far, compute max profit at each step

def max_profit(prices):
    pass


run_tests(max_profit, [
    (([7,1,5,3,6,4],), 5),
    (([7,6,4,3,1],),   0),
    (([2,4,1],),        2),
    (([1],),            0),
])

File: 06-arrays-sorting/02_buy_sell_stock.c
#include "../test_runner.h"

int max_profit(int* prices, int n) {
    return -1;
}

int main() {
    TEST_INT("basic",     max_profit(ARR(7,1,5,3,6,4), 6), 5);
    TEST_INT("decreasing",max_profit(ARR(7,6,4,3,1), 5),   0);
    TEST_INT("small",     max_profit(ARR(2,4,1), 3),        2);
    TEST_INT("single",    max_profit(ARR(1), 1),             0);
    TEST_SUMMARY();
}
